docs/implementing-cards.md

```diff
@@ -1,388 +1,570 @@
-this.persistentEffect({
-    match: this,
-    effect: AbilityDsl.effects.modifyMilitarySkill(() => this.getTwiceOpponentsClaimedRings())
-}
-```
+Static upgrade stat bonuses from the printed upgrade values are automatically included in combat calculations for the attached unit.
 
-#### Attachment-based effects
+##### Effects targeting attached card
 
-A `whileAttached` method is provided to define persistent effects that are applied to the card an attachment is attached. These effects remain as long as the card is attached to its parent and the attachment has not been blanked.
+Since most upgrade abilities target the attached card, we have helper methods available to declare such abilities succintly.
 
+Most upgrades say that the attached unit gains a triggered ability:
 ```javascript
-// Attached character gains Pride.
-this.whileAttached({
-    effect: AbilityDsl.effects.addKeyword('pride')
+// Attached character gains ability 'On Attack: Exhaust the defender'
+this.addGainTriggeredAbilityTargetingAttached({
+    title: 'Exhaust the defender on attack',
+    // note here that context.source refers to the attached unit card, not the upgrade itself
+    when: { onAttackDeclared: (event, context) => event.attack.attacker === context.source },
+    targetResolver: {
+        cardCondition: (card, context) => card === context.event.attack.target,
+        immediateEffect: AbilityHelper.immediateEffects.exhaust()
+    }
 });
 ```
 
-If the effect has an additional requirement, an optional `match` function can be passed in.
+It is also common for an upgrade to grant a keyword to the attached:
 ```javascript
-// If attached character is unicorn, they gain +1M.
-this.whileAttached({
-    match: card => card.isFaction('unicorn'),
-    effect: AbilityDsl.effects.modifyMilitarySkill(1)
+// Attached character gains keyword 'Restore 2'
+this.addGainKeywordTargetingAttached({
+    keyword: KeywordName.Restore,
+    amount: 2
 });
 ```
 
-#### Applying multiple effects at once
-As a shorthand, it is possible to pass an array into the `effect` property to apply multiple effects that have the same conditions / matching functions.
-
+In some rare cases an upgrade's ability targets the attached card without giving it any new abilities
 ```javascript
-// This character gets +1M and +1P while you are less honorable than an opponent..
-this.persistentEffect({
-    condition: () => this.isLessHonorableThanOpponent(),
-    match: this,
-    effect: [
-        AbilityDsl.effects.modifyMilitarySkill(1),
-        AbilityDsl.effects.modifyPoliticalSkill(1)
-    ]
+// Entrenched ability
+this.addConstantAbilityTargetingAttached({
+    title: 'Attached unit cannot attack bases',
+    ongoingEffect: AbilityHelper.ongoingEffects.cannotAttackBase(),
 });
 ```
 
-#### Applying effects to cards which aren't in play
+#### **IGNORE THIS SECTION, STILL WIP**: ~~Applying effects to cards which aren't in play~~
```

Comment:
Can we update all Ignore this, WIP and TBD type notes in here to all be TODO for future finding and referencing?